sys/exec.c

```diff
@@ -1,21 +1,162 @@
+        return -1;
+    }
+    // Ignore version and os abi field
+    // Check file type
+    if(eh->e_type != ET_EXEC){
+        kprintf("[exec] Exec failed: ELF is not an executable\n");
+        return -1;
+    }
+    // Check machine architecture field
+    if(eh->e_machine != EM_MIPS){
+        kprintf("[exec] Exec failed: ELF target architecture is not MIPS\n");
+        return -1;
+    }
+
+    // Take note of the entry point
+    kprintf("[exec] Entry point will be at 0x%08x.\n",
+            (unsigned int)eh->e_entry);
+
+    // Ensure prog header size compliance
+    if(eh->e_phentsize != sizeof(Elf32_Phdr)/sizeof(uint8_t)){
+        kprintf("[exec] Exec failed: ELF uses non-standard program header size\n");
+        return -1;
+    }
+
+    // TODO: Take care of assigning thread/process ASID
+    vm_map_t* vmap = vm_map_new(PMAP_USER, 123);
+    // Note: we do not claim ownership of the map
+    set_active_vm_map(vmap);
+
+    // Iterate over prog headers
+    kprintf("[exec] ELF has %d program headers\n", eh->e_phnum);
+
+    const Elf32_Phdr* phs = (Elf32_Phdr*)(elf_image + eh->e_phoff);
+    for(uint8_t i = 0; i < eh->e_phnum; i++){
+        const Elf32_Phdr* ph = phs + i;
+        switch(ph->p_type){
+        case PT_NULL:
+        case PT_NOTE:
+        case PT_PHDR:
+        default:
+            // Ignore the section.
+            break;
+        case PT_DYNAMIC:
+        case PT_INTERP:
+            kprintf("[exec] Exec failed: ELF file requests dynamic linking"
+                    "by providing a PT_DYNAMIC and/or PT_INTERP segment.\n");
+            return -1;
+        case PT_SHLIB:
+            kprintf("[exec] Exec failed: ELF file contains a PT_SHLIB segment\n");
+            return -1;
+        case PT_LOAD:
+            kprintf("[exec] Processing a PT_LOAD segment: VirtAddr = %p, "
+                    "Offset = 0x%08x, FileSiz = 0x%08x, MemSiz = 0x%08x, "
+                    "Flags = %d\n", (void*)ph->p_vaddr, (unsigned int)ph->p_offset,
+                    (unsigned int)ph->p_filesz, (unsigned int)ph->p_memsz,
+                    (unsigned int)ph->p_flags);
+            if(ph->p_vaddr % PAGESIZE){
+                kprintf("[exec] Exec failed: Segment p_vaddr is not page alligned\n");
+                return -1;
+            }
+            vm_addr_t start = ph->p_vaddr;
+            vm_addr_t end   = roundup(ph->p_vaddr + ph->p_memsz, PAGESIZE);
+            // TODO: What if segments overlap?
+            // Temporarily permissive protection.
+            vm_map_entry_t* segment = vm_map_add_entry(vmap, start, end,
+                                    VM_PROT_READ|VM_PROT_WRITE|VM_PROT_EXEC);
+            // Allocate pages backing this segment.
+            segment->object = default_pager->pgr_alloc();
+            // Copy data into the segment
+            memcpy((uint8_t*)start, elf_image + ph->p_offset, ph->p_filesz);
+            // Zero the rest
+            if(ph->p_filesz > ph->p_memsz){
```

Comment:
I believe this is backwards. Memory at the end needs to be zeroed if `memsz` is larger than `filesz`.